pkg/fleet-manager/application/rollout_helper.go

```diff
@@ -0,0 +1,459 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	applicationapi "kurator.dev/kurator/pkg/apis/apps/v1alpha1"
+	fleetapi "kurator.dev/kurator/pkg/apis/fleet/v1alpha1"
+	fleetmanager "kurator.dev/kurator/pkg/fleet-manager"
+	"kurator.dev/kurator/pkg/fleet-manager/plugin"
+)
+
+const (
+	RolloutLabel = "kurator.dev/rollout"
+)
+
+func (a *ApplicationManager) fetchRolloutClusters(ctx context.Context,
+	app *applicationapi.Application,
+	kubeClient client.Client,
+	originFleet *fleetapi.Fleet,
+	syncPolicy *applicationapi.ApplicationSyncPolicy) (map[fleetmanager.ClusterKey]*fleetmanager.FleetCluster, error) {
+
+	log := ctrl.LoggerFrom(ctx)
+	destination := getPolicyDestination(app, syncPolicy)
+	ClusterInterfaceList, result, err := a.fetchFleetClusterList(ctx, originFleet, destination.ClusterSelector)
+	if err != nil || result.RequeueAfter > 0 {
+		return nil, err
+	}
+
+	fleetclusters := make(map[fleetmanager.ClusterKey]*fleetmanager.FleetCluster, len(ClusterInterfaceList))
+	for _, cluster := range ClusterInterfaceList {
+		kclient, err := fleetmanager.ClientForCluster(kubeClient, originFleet.Namespace, cluster)
+		if err != nil {
+			return nil, err
+		}
+
+		kind := cluster.GetObject().GetObjectKind().GroupVersionKind().Kind
+		fleetclusters[fleetmanager.ClusterKey{Kind: kind, Name: cluster.GetObject().GetName()}] = &fleetmanager.FleetCluster{
+			Secret:    cluster.GetSecretName(),
+			SecretKey: cluster.GetSecretKey(),
+			Client:    kclient,
+		}
+	}
+	log.Info("Successful to fetch destination clusters for Rollout")
+	return fleetclusters, nil
+}
+
+func (a *ApplicationManager) syncRolloutPolicyForCluster(ctx context.Context,
+	rolloutPolicy *applicationapi.RolloutConfig,
+	destinationClusters map[fleetmanager.ClusterKey]*fleetmanager.FleetCluster,
+	policyName string) (ctrl.Result, error) {
+
+	log := ctrl.LoggerFrom(ctx)
+
+	// check some conflicting fields
+	if result, err := checkRolloutConflict(ctx, rolloutPolicy); err != nil {
+		return result, fmt.Errorf("failed to start rollout because: %s", err)
+	}
+
+	label := map[string]string{
+		RolloutLabel: policyName,
+	}
+
+	for clusterKey, fleetCluster := range destinationClusters {
+		// If the workload is a deployment, get its information to create the HPA.
+		if rolloutPolicy.Workload.Kind == "Deployment" {
+			deployment := &appsv1.Deployment{}
+			if err := getResourceFromClusterClient(ctx, rolloutPolicy.Workload.Namespace, rolloutPolicy.Workload.Name, *fleetCluster, deployment); err != nil {
+				if apierrors.IsNotFound(err) {
+					return ctrl.Result{}, nil
+				}
+				return ctrl.Result{}, errors.Wrapf(err, "failed to get workload %s in %s", rolloutPolicy.Workload.Name, rolloutPolicy.Workload.Namespace)
+			}
+
+			hpa := createHPAConfig(rolloutPolicy.Workload, deployment)
```

Comment:
As we discussed before, i think we agreed on not setting HPA by default

Reply:
In Flagger, if you want to create a canary for a deploy to do a roloout, you need a HPA.
If create a canary for a deploy that dosent have an HPA, flagger will report an error:
![image](https://github.com/kurator-dev/kurator/assets/140870529/c0f6d8f1-6bad-4b60-8955-9a5379962684)

Reply:
It's True. If don't set autoScalerRef in canary, we dont have to set HPA.